sovtokenfees/sovtokenfees/fees_authorizer.py

```diff
@@ -12,31 +15,79 @@
+    def __init__(self,
+                 fees_req_handler: StaticFeesReqHandler,
+                 config_state: PruningState,
+                 utxo_cache: UTXOCache):
         super().__init__()
         self.fees_req_handler = fees_req_handler
         self.config_state = config_state
+        self.utxo_cache = utxo_cache
         self.state_serializer = StaticFeesReqHandler.state_serializer
 
-    def _get_fees_alias_from_constraint(self, auth_constaint: AuthConstraint):
-        if auth_constaint.metadata:
-            if FEES_FIELD_NAME in auth_constaint.metadata:
-                return auth_constaint.metadata[FEES_FIELD_NAME]
-
-    def _can_pay_fees(self, request, required_fees):
+    def can_pay_fees(self, request, required_fees):
         try:
-            self.fees_req_handler.can_pay_fees(request, required_fees)
+            self._can_pay_fees(request, required_fees)
         except Exception as e:
             return False, str(e)
+
         return True, ''
 
+    def _can_pay_fees(self, request, required_fees):
+
+        if request.operation[TXN_TYPE] == XFER_PUBLIC:
+            # Fees in XFER_PUBLIC is part of operation[INPUTS]
+            inputs = request.operation[INPUTS]
+            outputs = request.operation[OUTPUTS]
+            self._validate_fees_can_pay(request, inputs, outputs, required_fees)
+            self.fees_req_handler.deducted_fees_xfer[request.key] = required_fees
+        else:
+            inputs = request.fees[0]
+            outputs = StaticFeesReqHandler.get_change_for_fees(request)
+            self._validate_fees_can_pay(request, inputs, outputs, required_fees)
+
+    def _validate_fees_can_pay(self, request, inputs, outputs, required_fees):
+        """
+        Calculate and verify that inputs and outputs for fees can both be paid and change is properly specified
+
+        This function ASSUMES that validation of the fees for the request has already been done.
+
+        :param request:
+        :param required_fees:
+        :return:
+        """
+
+        try:
+            sum_inputs = self.utxo_cache.sum_inputs(inputs, is_committed=False)
+        except UTXOError as ex:
+            raise InvalidFundsError(request.identifier, request.reqId, "{}".format(ex))
+        except Exception as ex:
+            error = 'Exception {} while processing inputs/outputs'.format(ex)
+            raise UnauthorizedClientRequest(request.identifier, request.reqId, error)
+        else:
+            change_amount = sum([a[AMOUNT] for a in outputs])
+            expected_amount = change_amount + required_fees
+            TokenReqHandler.validate_given_inputs_outputs(
+                sum_inputs,
+                change_amount,
+                expected_amount,
+                request,
+                'fees: {}'.format(required_fees)
+            )
     def _calculate_fees_from_req(self, request):
         return self.fees_req_handler.calculate_fees_from_req(request)
```

Comment:
I think `calculate_fees_from_req` needs to be moved to `FeesAuthorizer` as well

Reply:
We may make `calculate_fees_from_req ` static there. After this we can remove `fees_req_handler ` from this class at all.